src/v/cluster/tests/partition_balancer_simulator_test.cc

```diff
@@ -0,0 +1,273 @@
+          ntp);
+        BOOST_REQUIRE(cur_assignment);
+
+        cluster::finish_moving_partition_replicas_cmd cmd{
+          ntp, cur_assignment->replicas};
+
+        _workers.dispatch_topic_command(std::move(cmd));
+
+        size_t size = max_partition_size(ntp);
+
+        absl::flat_hash_set<model::node_id> cur_replicas;
+        for (const auto& bs : cur_assignment->replicas) {
+            cur_replicas.insert(bs.node_id);
+        }
+        for (auto& [id, node] : _nodes) {
+            if (cur_replicas.contains(id)) {
+                size_t old = std::exchange(node.replica_sizes[ntp], size);
+                node.used += size - old;
+            } else {
+                auto it = node.replica_sizes.find(ntp);
+                if (it != node.replica_sizes.end()) {
+                    node.used -= it->second;
+                    node.replica_sizes.erase(it);
+                }
+            }
+        }
+    }
+
+    size_t max_partition_size(const model::ntp& ntp) const {
+        size_t size = 0;
+        for (const auto& [id, node] : _nodes) {
+            auto it = node.replica_sizes.find(ntp);
+            if (it != node.replica_sizes.end()) {
+                size = std::max(size, it->second);
+            }
+        }
+        return size;
+    }
+
+    struct node_state {
+        model::node_id id;
+        size_t total = 0;
+        size_t used = 0;
+        absl::flat_hash_map<model::ntp, size_t> replica_sizes;
+
+        cluster::node_health_report get_health_report() const {
+            cluster::node_health_report report;
+            storage::disk node_disk{.free = total - used, .total = total};
+            report.id = id;
+            report.local_state.set_disk(node_disk);
+
+            absl::flat_hash_map<
+              model::topic_namespace,
+              ss::chunked_fifo<cluster::partition_status>>
+              topic2partitions;
+            for (const auto& [ntp, size] : replica_sizes) {
+                topic2partitions[model::topic_namespace(ntp.ns, ntp.tp.topic)]
+                  .push_back(cluster::partition_status{
+                    .id = ntp.tp.partition, .size_bytes = size});
+            }
+
+            for (auto& [topic, partitions] : topic2partitions) {
+                report.topics.push_back(
+                  cluster::topic_status(topic, std::move(partitions)));
+            }
+
+            return report;
+        }
+    };
+
+    absl::btree_map<model::node_id, node_state> _nodes;
+    controller_workers _workers;
+};
+
+FIXTURE_TEST(test_decommission, partition_balancer_sim_fixture) {
+    for (size_t i = 0; i < 4; ++i) {
+        add_node(model::node_id{i}, 100_GiB);
+    }
+    add_topic("mytopic", 500, 3, 100_MiB);
+    set_decommissioning(model::node_id{0});
+
+    print_state();
+    for (size_t i = 0; i < 1000; ++i) {
+        if (nodes().at(model::node_id{0}).replica_sizes.empty()) {
+            break;
+        }
+        tick();
+        print_state();
+    }
+
```

Comment:
this is pretty cool, looks like the intention of the simulator is to test the actual planning algorithm (and not the backend scheduler around it)..

we should probably loop these tests with various combinations of dimensions like node_count/topics/partitions/free disk availability/constraints/cores etc..  and have some final state validation based on some golden spec (like how many ticks would it take to reach that state).

Reply:
Yeah, there are a lot of scenarios to experiment with!